Reference/Common-Pitfalls/index-v9.md

```diff
@@ -0,0 +1,426 @@
+        {
+            // UmbracoContext
+            UmbracoContext.Content.GetById(1234);
+        }
+
+        return Ok();
+    }
+}
+```
+
+So next time you are using a singleton pattern or a static, think "Why am I doing this?", "I'm using Dependency Injection, I should be injecting this instance into my class."
+
+## Static references to scoped instances (such as `UmbracoHelper`)
+
+__Example 1:__
+
+```csharp
+public class BadApiController : UmbracoApiController
+{
+    // Note that this is static, that's bad
+    private static UmbracoHelper _umbracoHelper;
+    
+    public BadApiController(IUmbracoHelperAccessor umbracoHelperAccessor)
+    {
+        // Don't do this, this is bad
+        if (_umbracoHelper is null)
+        {
+            umbracoHelperAccessor.TryGetUmbracoHelper(out UmbracoHelper umbracoHelper);
+            _umbracoHelper = umbracoHelper;
+        }
+    }
+}
+```
+
+This practice can cause memory leaks along with inconsistent data results when using this `_umbracoHelper` instance.
+
+__Why?__
+
+It's important to understand the difference between an object that has a Request based scope/lifespan and
+an object that has an Singleton/Application based scope/lifespan ... here's the gist:
+
+* Application scope - if an object has an singleton/application scope/lifespan, that means that this single object
+instance will exist for the lifetime of the application. The single instance will be shared by every thread that
+accesses it. Static variables will always be application lifespan.
+* Request scope - The web world is made up of requests and each request has its own thread. When an object is in the scope of a Request it only survives as long as the web request survives. At the end of the web request, it may either be disposed or cleared from memory by the garbage collector. Request scoped object instances are not accessed by every other thread in the application - __unless you do something like the above!__
+
+An example of a request scoped instance is the `HttpContext`. This object exists for a single request and it definitely cannot be shared between other threads and especially not other request threads. This is because it is where the security information for a given user is stored! The `UmbracoContext` is also a request scoped object - in fact it relies directly on an instance of `HttpContext`. The `UmbracoHelper` is request scoped as well.
+
+So... in the above example, the `UmbracoHelper` which has a request scoped lifetime, will now be statically assigned to a variable, which means that this particular request scoped object is now bound to an Application scope lifetime and will never go away. This could mean that under certain circumstances that an entire Umbraco
+cache copy is stuck in memory, or that the `Security` property of the context will be accessed by multiple threads but this now contains the security information for a user for another request!
```

Comment:
```suggestion
An example of a request-scoped instance is the `HttpContext`. This object exists for a single request and it definitely cannot be shared between other threads and especially not other request threads. This is because it is where the security information for a given user is stored! The `UmbracoContext` is also a request-scoped object - in fact, it relies directly on an instance of `HttpContext`. The `UmbracoHelper` is request-scoped as well.

So... in the above example, the `UmbracoHelper` which has a request-scoped lifetime, will now be statically assigned to a variable, which means that this particular request-scoped object is now bound to an Application scope lifetime and will never go away. This could mean that under certain circumstances that an entire Umbraco
cache copy is stuck in memory, or that the `Security` property of the context will be accessed by multiple threads but this now contains the security information for a user for another request!
```